pkg/storage/opensearch.go

```diff
@@ -0,0 +1,385 @@
+	config := opensearchapi.Config{
+		Client: opensearch.Config{
+			Addresses: []string{url},
+		},
+	}
+
+	// Add basic auth if credentials are provided
+	if username != "" && password != "" {
+		config.Client.Username = username
+		config.Client.Password = password
+	}
+
+	var err error
+	os.osClient, err = opensearchapi.NewClient(config)
+	if err != nil {
+		// TODO - Add instrumentation here for failed opensearch connection
+		panic(err)
+	}
+}
+
+// osFieldMapping is an alias to the shared CADFFieldMapping for consistency.
+// The field mapping is defined in util.go to ensure consistency across storage backends.
+var osFieldMapping = CADFFieldMapping
+
+// buildBoolQuery constructs a bool query JSON string from filters
+func buildBoolQuery(filter *EventFilter) map[string]any {
+	boolQuery := map[string]any{
+		"bool": map[string]any{
+			"must":     []any{},
+			"filter":   []any{},
+			"must_not": []any{},
+		},
+	}
+
+	boolClause := boolQuery["bool"].(map[string]any)
+
+	// Helper to add filter or negation
+	addFilter := func(value, fieldName string) {
+		if strings.HasPrefix(value, "!") {
+			// Negation: add to must_not
+			value = value[1:]
+			boolClause["must_not"] = append(boolClause["must_not"].([]any), map[string]any{
+				"term": map[string]any{fieldName: value},
+			})
+		} else {
+			// Normal filter: add to filter
+			boolClause["filter"] = append(boolClause["filter"].([]any), map[string]any{
+				"term": map[string]any{fieldName: value},
+			})
+		}
+	}
+
+	if filter.ObserverType != "" {
+		addFilter(filter.ObserverType, osFieldMapping["observer_type"])
+	}
+	if filter.TargetType != "" {
+		addFilter(filter.TargetType, osFieldMapping["target_type"])
+	}
+	if filter.TargetID != "" {
+		addFilter(filter.TargetID, osFieldMapping["target_id"])
+	}
+	if filter.InitiatorType != "" {
+		addFilter(filter.InitiatorType, osFieldMapping["initiator_type"])
+	}
+	if filter.InitiatorID != "" {
+		addFilter(filter.InitiatorID, osFieldMapping["initiator_id"])
+	}
+	if filter.InitiatorName != "" {
+		addFilter(filter.InitiatorName, osFieldMapping["initiator_name"])
+	}
+	if filter.Action != "" {
+		addFilter(filter.Action, osFieldMapping["action"])
+	}
+	if filter.Outcome != "" {
+		addFilter(filter.Outcome, osFieldMapping["outcome"])
+	}
+	if filter.RequestPath != "" {
+		addFilter(filter.RequestPath, osFieldMapping["request_path"])
+	}
+
+	// Time range filters
+	if len(filter.Time) > 0 {
+		rangeQuery := map[string]any{osFieldMapping["time"]: map[string]any{}}
+		timeRange := rangeQuery[osFieldMapping["time"]].(map[string]any)
+
+		for key, value := range filter.Time {
+			switch key {
+			case "lt":
+				timeRange["lt"] = value
+			case "lte":
+				timeRange["lte"] = value
+			case "gt":
+				timeRange["gt"] = value
+			case "gte":
+				timeRange["gte"] = value
+			}
+		}
+
+		boolClause["filter"] = append(boolClause["filter"].([]any), map[string]any{"range": rangeQuery})
+	}
+
+	// Full-text search
+	if filter.Search != "" {
```

Comment:
A bit of a general observation. The data validation for all the "filter" fields (Offset, Limit, Time, MaxDepth etc.) is thorough with boundary checks and whitelisted values. However, for the free-text fields there is a lack of validation. What are your thoughts on this? Is this not a prio/concern from a security and integrity perspective?

I am primarily looking at the `filter.Search` field. But technically any unbounded string field ([here](https://github.com/SAP-cloud-infrastructure/hermez/blob/75a915e4afe83299671a374724ae63c255cce2e8/pkg/hermes/events.go#L41C1-L51C1)) are also free-text fields.

Reply:
And actually, I just confirmed pretty quickly that using filters to do index scoping for multi tenant I can hack around quite easily. Ugh. This throws our whole idea of a single index into quesiton. Thank you for giving me the perspective to see this.